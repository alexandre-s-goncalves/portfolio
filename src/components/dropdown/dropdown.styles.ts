import styled from 'styled-components';
import {colors} from 'resources/colors';
import {margin} from 'resources/margins';
import {Text} from 'components/ui/text';
import {Icon} from 'components/icon';
import GlobeSVG from 'assets/icons/iGlobe.svg';

interface ExtraProps {
  $open?: boolean;
  $active?: boolean;
  $clicked?: boolean;
  $rotate?: boolean;
  $theme?: boolean;
}

export const ContainerOption = styled.div`
  display: flex;
  align-items: center;
  gap: ${margin.xsmall}px;
`;

export const DropdownContainer = styled.div`
  position: relative;
  margin: ${margin.xsmall}px;
`;

export const Flag = styled(Icon).attrs<ExtraProps>(({$clicked}) => ({
  height: margin.medium,
  width: margin.medium,
  cursor: $clicked ? 'default' : 'pointer',
}))``;

export const IconGlobe = styled(Icon).attrs<ExtraProps>(({$clicked, $theme}) => ({
  icon: GlobeSVG,
  color: $theme ? colors.background01 : colors.black,
  height: margin.medium,
  width: margin.medium,
  cursor: $clicked ? 'default' : 'pointer',
}))``;

export const Menu = styled.ul<ExtraProps>`
  position: absolute;
  background: ${({$theme}) => ($theme ? colors.background05 : colors.background01)};
  border: 1px solid ${({$theme}) => ($theme ? colors.charcoal : colors.neutralLight)};
  box-shadow: 0 ${margin.xsmall}px ${margin.small}px rgba(0, 0, 0, 0.2);
  border-radius: ${margin.xsmall}px;
  color: ${colors.cadetGrey};
  top: ${margin.xxlarge}px;
  left: 100%;
  min-width: 150px;
  transform: translateX(-50%);
  opacity: ${({$open}) => ($open ? 1 : 0)};
  transition: opacity 0.2s;
  display: ${({$open}) => ($open ? 'block' : 'none')};
  z-index: 1;
`;

export const MenuItem = styled.li<ExtraProps>`
  display: flex;
  padding: ${margin.xsmall}px;
  margin: ${margin.xxsmall}px;
  border-radius: ${margin.xsmall}px;
  cursor: ${({$active}) => ($active ? 'default' : 'pointer')};
  background: ${({$active, $theme}) =>
    $theme
      ? $active
        ? colors.raisinBlack
        : 'transparent'
      : $active
        ? colors.background03
        : 'transparent'};

  &:hover {
    background: ${({$theme}) => ($theme ? colors.raisinBlack : colors.background03)};
  }
`;

export const Select = styled.div<ExtraProps>`
  background: ${({$theme}) => ($theme ? colors.background05 : colors.background01)};
  display: flex;
  justify-content: center;
  align-items: center;
  width: ${margin.xlarge}px;
  border-radius: ${margin.xsmall}px;
  padding: 8px 12px;
  cursor: ${({$clicked}) => ($clicked ? 'default' : 'pointer')};

  &:hover {
    background: ${({$theme, $clicked}) =>
      $clicked
        ? $theme
          ? colors.background05
          : colors.background01
        : $theme
          ? colors.gunmetal
          : colors.background03};
  }
`;

export const Selected = styled.span`
  display: flex;
  align-items: center;
  gap: ${margin.xsmall}px;
`;

export const Texto = styled(Text).attrs<ExtraProps>(({$theme, $active, $clicked}) => ({
  color: $theme ? colors.black : colors.background01,
  cursor: $active || $clicked ? 'default' : 'pointer',
}))``;
