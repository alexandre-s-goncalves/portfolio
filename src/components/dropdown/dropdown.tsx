import React, {useState, useEffect, useMemo, useRef} from 'react';
import {useTranslation} from 'react-i18next';
import {languages, namespaces} from 'utils/i18n/i18n.constants';
import {useTheme} from 'context/ThemeContext';
import FlagBrasilSVG from 'assets/icons/iFlag-Brasil-active.svg';
import FlagEUASVG from 'assets/icons/iFlag-USA-active.svg';
import FlagFranceSVG from 'assets/icons/iFlag-France-active.svg';
import * as S from './dropdown.styles';

interface IOption {
  testId: string;
  icon: React.ElementType;
  label: () => string;
  abbreviation: string;
}

export const Dropdown = () => {
  const {theme} = useTheme();
  const {t, i18n} = useTranslation(namespaces.components.dropdown);
  const [isOpen, setIsOpen] = useState(false);
  const [selectedOption, setSelectedOption] = useState<IOption | null>(null);
  const [activeOption, setActiveOption] = useState<IOption | null>(null);
  const dropdownRef = useRef<HTMLDivElement>(null);

  const options = useMemo(
    () => [
      {
        testId: 'idioma-portugues',
        abbreviation: languages.pt,
        icon: FlagBrasilSVG,
        label: () => t('portuguese'),
      },
      {
        testId: 'idioma-ingles',
        abbreviation: languages.en,
        icon: FlagEUASVG,
        label: () => t('english'),
      },
      {
        testId: 'idioma-frances',
        abbreviation: languages.fr,
        icon: FlagFranceSVG,
        label: () => t('french'),
      },
    ],
    [t],
  );

  useEffect(() => {
    const resolvedLanguage = i18n.resolvedLanguage;

    const initialOption = options.find(option => {
      if (resolvedLanguage === 'pt') return option.testId === 'idioma-portugues';
      if (resolvedLanguage === 'en') return option.testId === 'idioma-ingles';
      if (resolvedLanguage === 'fr') return option.testId === 'idioma-frances';
      return null;
    });

    if (initialOption) {
      setSelectedOption(initialOption);
      setActiveOption(initialOption);
    }
  }, [i18n.resolvedLanguage, options]);

  useEffect(() => {
    const handleClickOutside = (event: MouseEvent) => {
      if (dropdownRef.current && !dropdownRef.current.contains(event.target as Node)) {
        setIsOpen(false);
      }
    };

    globalThis.addEventListener('mousedown', handleClickOutside);
    return () => {
      globalThis.removeEventListener('mousedown', handleClickOutside);
    };
  }, []);

  const handleSelectClick = () => {
    if (!isOpen) {
      setIsOpen(true);
    }
  };

  const handleOptionClick = (option: IOption) => {
    setSelectedOption(option);
    setActiveOption(option);
    i18n.changeLanguage(option.abbreviation);
    setIsOpen(false);
  };

  return (
    <S.DropdownContainer ref={dropdownRef}>
      <S.Select
        data-testid="dropdown-select"
        $theme={theme}
        $clicked={isOpen}
        onClick={handleSelectClick}>
        <S.Selected>
          {selectedOption && (
            <S.IconGlobe
              data-testid={`selected-${selectedOption.testId}`}
              $clicked={isOpen}
              $theme={theme}
            />
          )}
          <S.Texto $theme={!theme} $clicked={isOpen}>
            {selectedOption ? selectedOption.abbreviation.toUpperCase() : t('select')}
          </S.Texto>
        </S.Selected>
      </S.Select>
      <S.Menu data-testid="dropdown-menu" $theme={theme} $open={isOpen}>
        {options.map(option => (
          <S.MenuItem
            key={option.testId}
            $active={activeOption === option}
            $theme={theme}
            onClick={() => handleOptionClick(option)}
            data-testid={option.testId}>
            <S.ContainerOption>
              <S.Flag icon={option.icon} data-testid={`flag-${option.testId}`} />
              <S.Texto $theme={!theme} $active={activeOption === option}>
                {option.label()}
              </S.Texto>
            </S.ContainerOption>
          </S.MenuItem>
        ))}
      </S.Menu>
    </S.DropdownContainer>
  );
};
