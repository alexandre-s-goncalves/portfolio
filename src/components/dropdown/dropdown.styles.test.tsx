import React from 'react';
import {render} from '@testing-library/react';
import '@testing-library/jest-dom';
import 'jest-styled-components';
import * as S from './dropdown.styles';

describe('dropdown.styles', () => {
  const Generic = () => <svg />;

  test('ContainerOption should match the snapshot', () => {
    const {container} = render(<S.ContainerOption />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('DropdownContainer should match the snapshot', () => {
    const {container} = render(<S.DropdownContainer />);
    expect(container.firstChild).toMatchSnapshot();
  });

  describe('Flag', () => {
    test('should match snapshot when clicked', () => {
      const {container} = render(<S.Flag icon={Generic} $clicked />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('should match snapshot when not clicked', () => {
      const {container} = render(<S.Flag icon={Generic} $clicked={false} />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('IconGlobe', () => {
    test('should match snapshot when clicked and themed', () => {
      const {container} = render(<S.IconGlobe $clicked $theme />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('should match snapshot when clicked and not themed', () => {
      const {container} = render(<S.IconGlobe $clicked $theme={false} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('should match snapshot when not clicked and themed', () => {
      const {container} = render(<S.IconGlobe $clicked={false} $theme />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('should match snapshot when not clicked and not themed', () => {
      const {container} = render(<S.IconGlobe $clicked={false} $theme={false} />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('Menu', () => {
    test('should match snapshot when open and themed', () => {
      const {container} = render(<S.Menu $open $theme />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('should match snapshot when open and not themed', () => {
      const {container} = render(<S.Menu $open $theme={false} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('should match snapshot when closed and themed', () => {
      const {container} = render(<S.Menu $open={false} $theme />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('should match snapshot when closed and not themed', () => {
      const {container} = render(<S.Menu $open={false} $theme={false} />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('MenuItem', () => {
    test('should match snapshot when active and themed', () => {
      const {container} = render(<S.MenuItem $active $theme />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('should match snapshot when active and not themed', () => {
      const {container} = render(<S.MenuItem $active $theme={false} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('should match snapshot when not active and themed', () => {
      const {container} = render(<S.MenuItem $active={false} $theme />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('should match snapshot when not active and not themed', () => {
      const {container} = render(<S.MenuItem $active={false} $theme={false} />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('Select', () => {
    test('should match snapshot when clicked and themed', () => {
      const {container} = render(<S.Select $clicked $theme />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('should match snapshot when clicked and not themed', () => {
      const {container} = render(<S.Select $clicked $theme={false} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('should match snapshot when not clicked and themed', () => {
      const {container} = render(<S.Select $clicked={false} $theme />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('should match snapshot when not clicked and not themed', () => {
      const {container} = render(<S.Select $clicked={false} $theme={false} />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  test('Selected should match snapshot', () => {
    const {container} = render(<S.Selected />);
    expect(container.firstChild).toMatchSnapshot();
  });

  describe('Texto', () => {
    test('should match snapshot when themed, active and clicked', () => {
      const {container} = render(<S.Texto $theme $active $clicked />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('should match snapshot when not themed, active and clicked', () => {
      const {container} = render(<S.Texto $theme={false} $active $clicked />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('should match snapshot when themed, not active and not clicked', () => {
      const {container} = render(<S.Texto $theme $active={false} $clicked={false} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('should match snapshot when not themed, not active and not clicked', () => {
      const {container} = render(<S.Texto $theme={false} $active={false} $clicked={false} />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
