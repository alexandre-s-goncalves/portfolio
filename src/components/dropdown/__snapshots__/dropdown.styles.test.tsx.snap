// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dropdown.styles ContainerOption should match the snapshot 1`] = `
.c0 {
  display: flex;
  align-items: center;
  gap: 8px;
}

<div
  class="c0"
/>
`;

exports[`dropdown.styles DropdownContainer should match the snapshot 1`] = `
.c0 {
  position: relative;
  margin: 8px;
}

<div
  class="c0"
/>
`;

exports[`dropdown.styles Flag should match snapshot when clicked 1`] = `<svg />`;

exports[`dropdown.styles Flag should match snapshot when not clicked 1`] = `<svg />`;

exports[`dropdown.styles IconGlobe should match snapshot when clicked and not themed 1`] = `
<svg
  class=""
  cursor="default"
  fill="#000000"
  height="24"
  style="min-width: 24px; min-height: 24px; max-width: 24px; max-height: 24px;"
  width="24"
/>
`;

exports[`dropdown.styles IconGlobe should match snapshot when clicked and themed 1`] = `
<svg
  class=""
  cursor="default"
  fill="#ffffff"
  height="24"
  style="min-width: 24px; min-height: 24px; max-width: 24px; max-height: 24px;"
  width="24"
/>
`;

exports[`dropdown.styles IconGlobe should match snapshot when not clicked and not themed 1`] = `
<svg
  class=""
  cursor="pointer"
  fill="#000000"
  height="24"
  style="min-width: 24px; min-height: 24px; max-width: 24px; max-height: 24px;"
  width="24"
/>
`;

exports[`dropdown.styles IconGlobe should match snapshot when not clicked and themed 1`] = `
<svg
  class=""
  cursor="pointer"
  fill="#ffffff"
  height="24"
  style="min-width: 24px; min-height: 24px; max-width: 24px; max-height: 24px;"
  width="24"
/>
`;

exports[`dropdown.styles Menu should match snapshot when closed and not themed 1`] = `
.c0 {
  position: absolute;
  background: #ffffff;
  border: 1px solid #D9DDDB;
  box-shadow: 0 8px 16px rgba(0, 0, 0, 0.2);
  border-radius: 8px;
  color: #9fa5b5;
  top: 64px;
  left: 100%;
  min-width: 150px;
  transform: translateX(-50%);
  opacity: 0;
  transition: opacity 0.2s;
  display: none;
  z-index: 1;
}

<ul
  class="c0"
/>
`;

exports[`dropdown.styles Menu should match snapshot when closed and themed 1`] = `
.c0 {
  position: absolute;
  background: #000000;
  border: 1px solid #363a43;
  box-shadow: 0 8px 16px rgba(0, 0, 0, 0.2);
  border-radius: 8px;
  color: #9fa5b5;
  top: 64px;
  left: 100%;
  min-width: 150px;
  transform: translateX(-50%);
  opacity: 0;
  transition: opacity 0.2s;
  display: none;
  z-index: 1;
}

<ul
  class="c0"
/>
`;

exports[`dropdown.styles Menu should match snapshot when open and not themed 1`] = `
.c0 {
  position: absolute;
  background: #ffffff;
  border: 1px solid #D9DDDB;
  box-shadow: 0 8px 16px rgba(0, 0, 0, 0.2);
  border-radius: 8px;
  color: #9fa5b5;
  top: 64px;
  left: 100%;
  min-width: 150px;
  transform: translateX(-50%);
  opacity: 1;
  transition: opacity 0.2s;
  display: block;
  z-index: 1;
}

<ul
  class="c0"
/>
`;

exports[`dropdown.styles Menu should match snapshot when open and themed 1`] = `
.c0 {
  position: absolute;
  background: #000000;
  border: 1px solid #363a43;
  box-shadow: 0 8px 16px rgba(0, 0, 0, 0.2);
  border-radius: 8px;
  color: #9fa5b5;
  top: 64px;
  left: 100%;
  min-width: 150px;
  transform: translateX(-50%);
  opacity: 1;
  transition: opacity 0.2s;
  display: block;
  z-index: 1;
}

<ul
  class="c0"
/>
`;

exports[`dropdown.styles MenuItem should match snapshot when active and not themed 1`] = `
.c0 {
  display: flex;
  padding: 8px;
  margin: 4px;
  border-radius: 8px;
  cursor: default;
  background: #F2F2F2;
}

.c0:hover {
  background: #F2F2F2;
}

<li
  class="c0"
/>
`;

exports[`dropdown.styles MenuItem should match snapshot when active and themed 1`] = `
.c0 {
  display: flex;
  padding: 8px;
  margin: 4px;
  border-radius: 8px;
  cursor: default;
  background: #23242a;
}

.c0:hover {
  background: #23242a;
}

<li
  class="c0"
/>
`;

exports[`dropdown.styles MenuItem should match snapshot when not active and not themed 1`] = `
.c0 {
  display: flex;
  padding: 8px;
  margin: 4px;
  border-radius: 8px;
  cursor: pointer;
  background: transparent;
}

.c0:hover {
  background: #F2F2F2;
}

<li
  class="c0"
/>
`;

exports[`dropdown.styles MenuItem should match snapshot when not active and themed 1`] = `
.c0 {
  display: flex;
  padding: 8px;
  margin: 4px;
  border-radius: 8px;
  cursor: pointer;
  background: transparent;
}

.c0:hover {
  background: #23242a;
}

<li
  class="c0"
/>
`;

exports[`dropdown.styles Select should match snapshot when clicked and not themed 1`] = `
.c0 {
  background: #ffffff;
  display: flex;
  justify-content: center;
  align-items: center;
  width: 48px;
  border: 1px solid transparent;
  border-radius: 8px;
  padding: 8px 12px;
  cursor: default;
}

.c0:hover {
  background: #ffffff;
}

<div
  class="c0"
/>
`;

exports[`dropdown.styles Select should match snapshot when clicked and themed 1`] = `
.c0 {
  background: #000000;
  display: flex;
  justify-content: center;
  align-items: center;
  width: 48px;
  border: 1px solid transparent;
  border-radius: 8px;
  padding: 8px 12px;
  cursor: default;
}

.c0:hover {
  background: #000000;
}

<div
  class="c0"
/>
`;

exports[`dropdown.styles Select should match snapshot when not clicked and not themed 1`] = `
.c0 {
  background: #ffffff;
  display: flex;
  justify-content: center;
  align-items: center;
  width: 48px;
  border: 1px solid #000000;
  border-radius: 8px;
  padding: 8px 12px;
  cursor: pointer;
}

.c0:hover {
  background: #F2F2F2;
}

<div
  class="c0"
/>
`;

exports[`dropdown.styles Select should match snapshot when not clicked and themed 1`] = `
.c0 {
  background: #000000;
  display: flex;
  justify-content: center;
  align-items: center;
  width: 48px;
  border: 1px solid #2a2f3b;
  border-radius: 8px;
  padding: 8px 12px;
  cursor: pointer;
}

.c0:hover {
  background: #323741;
}

<div
  class="c0"
/>
`;

exports[`dropdown.styles Selected should match snapshot 1`] = `
.c0 {
  display: flex;
  align-items: center;
  gap: 8px;
}

<span
  class="c0"
/>
`;

exports[`dropdown.styles Texto should match snapshot when not themed, active and clicked 1`] = `
.c0 {
  font-size: 12px;
  font-family: AvertaStd-Regular;
  color: #ffffff;
  margin-top: 0px;
  font-weight: normal;
  text-transform: none;
  cursor: default;
}

<p
  class="c0 "
/>
`;

exports[`dropdown.styles Texto should match snapshot when not themed, not active and not clicked 1`] = `
.c0 {
  font-size: 12px;
  font-family: AvertaStd-Regular;
  color: #ffffff;
  margin-top: 0px;
  font-weight: normal;
  text-transform: none;
  cursor: pointer;
}

<p
  class="c0 "
/>
`;

exports[`dropdown.styles Texto should match snapshot when themed, active and clicked 1`] = `
.c0 {
  font-size: 12px;
  font-family: AvertaStd-Regular;
  color: #000000;
  margin-top: 0px;
  font-weight: normal;
  text-transform: none;
  cursor: default;
}

<p
  class="c0 "
/>
`;

exports[`dropdown.styles Texto should match snapshot when themed, not active and not clicked 1`] = `
.c0 {
  font-size: 12px;
  font-family: AvertaStd-Regular;
  color: #000000;
  margin-top: 0px;
  font-weight: normal;
  text-transform: none;
  cursor: pointer;
}

<p
  class="c0 "
/>
`;
